src/saveload/afterload.cpp

```diff
@@ -3241,6 +3241,50 @@ bool AfterLoadGame()
 		}
 	}
 
+	if (IsSavegameVersionBefore(SLV_CARGO_MOVEMENT)) {
+		/* Find a reasonable source and travel distance for cargo in vehicles. */
```

Comment:
Why do all that logic to try (and possibly fail) to restore source location from station ids when you already have it in `source_xy`?

Reply:
`source_xy` is not always a valid value, when loading older savegames. As such, it is good to test some other methods too. But I do see I never test for source_xy, which might be good to see if that is a valid tile first, and try the others after.

Reply:
Weird, but it is currently used for cargo payments so maybe it's fine? Even if you want to fix that it may be better to put that conversion in a version that introduced the issue as it's not really related to the teleport fix. You need a sane source_xy for payment check anyway.